sotodlib/io/hkdb.py

```diff
@@ -0,0 +1,366 @@
+    items =[]
+    for p in paths:
+        hk_file, hk_frames = get_items_from_file(p)
+        items.append(hk_file)
+        items.extend(hk_frames)
+
+    with db.Session.begin() as sess:
+        sess.add_all(items)
+
+
+def get_all_hk_files(cfg: HkConfig):
+    """
+    Gets list of all hk files in specified root directory
+    """
+    hk_files = []
+    for root, _, files in os.walk(cfg.hk_root):
+        for file in files:
+            if file.endswith('.g3'):
+                hk_files.append(os.path.join(root, file))
+    return hk_files
+
+
+def index_all(cfg: Union[HkConfig, str], show_pb=True, files_per_batch=1):
+    """
+    Indexes all files in the specified root directory.
+    """
+    hkdb = HkDb(cfg)
+    all_files = get_all_hk_files()
+
+    with hkdb.Session.begin() as sess:
+        archived_paths = [path for path, in sess.query(HkFile.path).all()]
+        remaining_files = sorted(list(set(all_files) - set(archived_paths)))
+    
+    for i in trange(0, len(remaining_files), files_per_batch, disable=(not show_pb)):
+        index_files(remaining_files[i:i+files_per_batch], hkdb)
+
+
+#####################
+# HK Loading stuff
+#####################
+
+@dataclass
+class Field:
+    agent: str
+    feed: str
+    field: str
+
+    def __str__(self):
+        return f"{self.agent}.{self.feed}.{self.field}"
+    
+    def matches(self, other):
+        if self.agent != other.agent:
+            return False
+        if self.feed != other.feed and self.feed != '*' and other.feed != '*':
+            return False
+        if self.field != other.field and self.field != '*' and other.field != '*':
+            return False
+        return True
+    
+    @classmethod
+    def from_str(cls, s):
+        try:
+            agent, feed, field = s.split('.')
+        except Exception:
+            raise ValueError(f"Could not parse field: {s}")
+
+        return cls(agent, feed, field)
+
+@dataclass
+class LoadSpec:
+    """
+    HK loading specification
+
+    Args
+    -----
+    cfg : HkConfig
+        HkConfig object
+    fields : List[str]
+        List of field specifications to load. This can either be a field
+        descriptor, of the format ``agent.feed.field``, or an alias defined in
+        the config. Field descriptors can contain wildcards, for instance
+        ``agent.*.*`` will load all fields belonging to the specified agent.
+        ``agent.feed.*`` and ``agent.*.field`` will also work as expected.
+    start : float 
+        Starting time of data to load
+    end : float
+        Ending time of data to load
+    """
+    cfg: HkConfig
+    fields: List[str]
+    start: float
+    end: float
+
+    def __post_init__(self):
+        fs = []
+        for f in self.fields:
+            if f in self.cfg.aliases:
+                fs.append(Field.from_str(self.cfg.aliases[f]))
+            else:
+                fs.append(Field.from_str(f))
+        self.fields = fs
+
+
+class HkResult:
```

Comment:
can this just be an axis manager?

Reply:
I don't think this quite fits into an AxisManager so easily... because hk fields aren't cosampled, most of the benefits of AxisManager aren't really applicable, and I'm not sure how axis-managers handle long field names that aren't proper python identifiers, such as `cryo-ls372-lsa21yc.temperatures.Channel_02_T`.

I think it would be better to leave this as is, and have a separate function that transforms this into an AxisManager when re-sampling hk to be aligned with the 'samps' axis.

Reply:
Sanah did the setup for getting axismanagers where the fields are co-sampled and then co-sampling. I don't see a reason not to use it?

Reply:
I really dislike the idea of making this load fn dependent on all of the hk -> axis-manager conversion stuff that is in hk_utils...

I think there is enough subtlety in how fields are grouped and how the cosampling is done, that it should happen separately and not as a part of the loading function.